update TLS profile document of the Metal3 stack (#94)
Nordix/metal3-clusterapi-docs: security/tls-1.3.md

For Ironic, it should be noted that:

- IPA image serving from `httpd` port `6180` might be insecure due to BMC or

>   (i)PXE firmware limitations, the user has to enable TLS for both
>   iPXE and virtual media and in case of iPXE custom iPXE firmware building
>   is required. In case TLS is enabled for virtual media and iPXE, additional
>   TLS enabled ports will be opened in the form of `8083` and `8084`

Reviewer: ```suggestion
  (i)PXE firmware limitations, the user has to enable TLS for both
  iPXE and virtual media. In case of iPXE, custom iPXE firmware building
  is required. If TLS is enabled for virtual media and iPXE, additional
  TLS enabled ports will be opened in the form of `8083` and `8084`
```

Author: Thanks done.

  respectively.

> - The `node image server` that hosts the node images could use `HTTPS` but
>   that would require an IPA build process that injects the relevant

Reviewer: ```suggestion
- The `node image server` could use `HTTPS` but
  that would require an IPA build process that injects the relevant
```

Author: Thanks done.

  certificate to the IPA. In addition to IPA, Ironic would also need access to

>   the node image server certificates. When the certificates are supplied IPA

Reviewer 2: ```suggestion
  the node image server certificates. Once the certificates are supplied, both IPA and Ironic support TLS 1.3 ...
```

Author: Thanks, done.

  and Ironic supports TLS 1.3 during the `node image download` workflow,
  the process is facilitated via the third party request
  [library](https://pypi.org/project/requests/), same library is used by Ironic
  for `validating the node images`. There is no way to enable TLS support
  for the `node image server` in case `ironic-image` is used to provide
  the `node image server` functionality.

- Ironic external endpoints are secured via `httpd` (Apache) that handles TLS
  termination. Between the `http proxy` and Ironic/Inspector the communication

>   could go via UNIX socket or http depending configuration. Ironic and Ironic

Reviewer 2: ```suggestion
  could go via UNIX socket or http depending on configuration. Ironic and Ironic
```

Author: Thanks done.

For Ironic, it should be noted that:

> - Ironic external endpoints are secured, while pod-internal traffic is not.

Reviewer 3: Should this also be changed? If we don't use `http` for pod internal traffic, but sockets?

**etcd**:

For etcd, TLS 1.3 support was
[contributed by EST](https://github.com/etcd-io/etcd/pull/15156) and was added
to the 3.5.8 (April 13 2023). Etcd versions from `3.5.0` to `3.5.7` have
hardcoded TLS 1.2 version. Some versions of etcd `3.4.x` have support for
TLS 1.3 as they lack this hardcoding of TLS version, and using new enough
Golang enables TLS 1.3 for them. TLS 1.3 can be explicitly enabled starting
from `3.4.25` in case of `3.4.x` series are used or starting from `3.5.8` in
case `3.5.x` series is used.

> - Port `2379`: TLS 1.2

Reviewer 3: ```suggestion
- Port `2379`: TLS 1.2, TLS 1.3
```

> - Port `2380`: TLS 1.2

Reviewer 3: ```suggestion
- Port `2380`: TLS 1.2,TLS 1.3
```

**etcd**:

For etcd, TLS 1.3 support was
[contributed by EST](https://github.com/etcd-io/etcd/pull/15156) and was added
to the 3.5.8 (April 13 2023). Etcd versions from `3.5.0` to `3.5.7` have
hardcoded TLS 1.2 version. Some versions of etcd `3.4.x` have support for
TLS 1.3 as they lack this hardcoding of TLS version, and using new enough
Golang enables TLS 1.3 for them. TLS 1.3 can be explicitly enabled starting

> from `3.4.25` in case of `3.4.x` series are used or starting from `3.5.8` in

Reviewer 2: ```suggestion
from `3.4.25` in case of `3.4.x` series is used or starting from `3.5.8` in
```

TLS 1.3 is well supported in Metal3 ecosystem, with two exceptions:

- Ironic Python Agent (IPA), where the Oslo library is not
  supporting TLS 1.3 officially but because of implementation characteristics

>   of `oslo.service` and the nature of Python 3 (up to 3.12) TLS 1.3 and 1.2

Reviewer 3: ```suggestion
  of `oslo.service` and the nature of Python 3 (from 3.7 up to 3.12) TLS 1.2 and 1.3
```

  auto-negotiation is enabled by default but TLS 1.3 exclusive connection can't
  be enforced.

As of 29.1.2024 TLS 1.3 is supported everywhere in the whole stack, but there

> are API endpoints in Ironic (pod internal) and IPA where TLS 1.3 can't be

Reviewer 3: Same as above, which pod internal interfaces are not covered?

> exclusively enforced instead there is TLS 1.3 - 1.2 auto-negotiation.

Reviewer 3: ```suggestion
exclusively enforced instead there is TLS 1.2/1.3 auto-negotiation.
```

## Summary

TLS 1.3 is well supported in Metal3 ecosystem, with a few caveats
as of 29.1.2024:

- Ironic Python Agent (IPA), where the Oslo library is not
  supporting TLS 1.3 officially but because of implementation characteristics
  of `oslo.service` and the nature of Python 3 (from 3.7 up to 3.12) TLS 1.3
  and 1.2 auto-negotiation is enabled by default but TLS 1.3 exclusive
  connection can't be enforced.

- Ironic and Ironic Inspector internal ports are available via HTTP from
  within the pod or the host machine by default, but could be switched to
  utilize `UNIX sockets`.

- httpd node image server doesn't support https but that is not an inherent

>   limitation of either the Ironic, IPA or httpd only how ironic-image httpd

Reviewer 2: ```suggestion
  limitation of either the Ironic, IPA or httpd but only how ironic-image httpd
```

Author: Thanks done.
